File: src/features/typing/data/go/generate_parentheses.go
package main

func generateParenthesis(n int) []string {
	result := []string{}
	var backtrack func(current []byte, open, close int)
	backtrack = func(current []byte, open, close int) {
		if len(current) == 2*n {
			result = append(result, string(current))
			return
		}
		if open < n {
			next := append(append([]byte{}, current...), '(')
			backtrack(next, open+1, close)
		}
		if close < open {
			next := append(append([]byte{}, current...), ')')
			backtrack(next, open, close+1)
		}
	}
	backtrack([]byte{}, 0, 0)
	return result
}

File: src/features/typing/data/dart/generate_parentheses.dart
class Solution {
  List<String> generateParenthesis(int n) {
    final result = <String>[];
    void backtrack(String current, int open, int close) {
      if (current.length == 2 * n) {
        result.add(current);
        return;
      }
      if (open < n) {
        backtrack('$current(', open + 1, close);
      }
      if (close < open) {
        backtrack('$current)', open, close + 1);
      }
    }
    backtrack('', 0, 0);
    return result;
  }
}
